app/controllers/users/two_factor_authentication_controller.rb

```diff
@@ -56,7 +52,7 @@ def delivery_preference
     end
 
     def update_otp_delivery_preference_if_needed
-      return unless user_signed_in?
+      return unless user_signed_in? && @telephony_result&.success?
```

Comment:
Personally I really dislike compound statement in an `unless` because it makes it harder for me to read --- WDYT about turning it into an `if`?